### DISCUSSION TITLE
Discuss: A modular approach to patterns

### DISCUSSION THREAD

**Discussion by joostdecock** [freesewing/freesewing#1107] [Forum]
This is a todo from #774 

An idea was raised to make it possible to have a modular pattern system where you can combine sleeves, necklines, bodies and so on.

This issue is a place to discuss and perhaps more.

**Comment by joostdecock** [comment]
Stale. Closing

**Comment by jejacks0n** [comment]
I'm going to reopen this @joostdecock, for discussion, if that's alright?

I've had some thoughts about the overall approach we could take for this, and if you feel comfortable with me exploring some concepts here -- the thinking is that 2d boolean operations might give us a path forward for this.

**Comment by joostdecock** [comment]
Sure thing. We can also turn this into a discussion item if that's more appropriate. Whatever suits 👍

**Comment by jejacks0n** [comment]
It's surfaced that many pattern concepts might be handled by doing boolean operations -- notably, boolean subtraction.

Arm and neck holes are good examples of potentially benefitting from this concept. The shape of them must often be cohesive across multiple parts, and so this could be handled by simply subtracting a single arm hole shape from the front bodice part and the back bodice part.

Changing the shape of the arm hole would then be the same operation on multiple parts, and the arm hole formula can be extracted into an arm hole "module" (or shape) that can be more easily shared.

**Comment by jejacks0n** [comment]
> Sure thing. We can also turn this into a discussion item if that's more appropriate. Whatever suits 👍

Whichever you prefer @joostdecock!

**Comment by starfetch** [comment]
This sounds like a good starting point for using Discussions :)

**Reply by starfetch** [reply]
So I've taken the liberty to move it.

**Reply by jejacks0n** [reply]
🙏  thanks @starfetch!

**Comment by jejacks0n** [comment]
Ok, here's a writeup of my findings, and potential approaches for how we might want to approach enabling these kinds of concepts.

I evaluated the following projects:

1. 2d-polygon-boolean - https://github.com/tmpvar/2d-polygon-boolean - [no demo]
1. ClipperJS - https://github.com/Doodle3D/clipper-js - [demo](http://jsclipper.sourceforge.net/6.4.2.2/main_demo.html)
1. Paper.js - http://paperjs.org - [demo](http://paperjs.org/examples/boolean-operations/)

The first, 2d-polygon-boolean was pretty quickly eliminated because in the [white paper](https://www.inf.usi.ch/hormann/papers/Greiner.1998.ECO.pdf) it references being built around, it's explicitly stated that it uses polygons.

The second option, ClipperJS looked really promising, and it's a solid library with a great approach, but when executing boolean operations on shapes, it first converts them to polygons, and so is fairly limiting in terms of curves and bezier curves -- they lose resolution in converting to polygons, and we want to try to avoid this for patterns.

The third option is where I spent the most time, because it has the ability to do boolean operations on 2d shapes, leaving the curves intact. This means that patterns wouldn't loose any clarity on curves when calculating boolean operations, as the first two approaches suffer.

So, my recommendation from initial investigations would be to further explore Paper.js as a viable library enhancement.

**Comment by jejacks0n** [comment]
I played around a bit with Paper.js, and was able to generate a pretty quick (and dirty) mock of how it could be used to either drive the freesewing `Path`/`Point`/etc. core library classes, or potentially replace some of them over time.

I wanted a very simple mockup of the idea, so I added the following to my basic bodice block pattern (that's a work in progress). 

I added this javascript to the `index.js`, leaving the current export logic in place:

```javascript
// existing code...

import paper from 'paper'

paper.setup([300, 300]);

const bodiceFrontSource = new paper.Path({ segments: [[0, 20], [220, 0], [140, 250]], closed: true });
const armHoleSource = new paper.Path.Circle({ center: [100, 150], radius: 30 })

const bodiceFront = bodiceFrontSource.subtract(armHoleSource);
bodiceFront.fillColor = 'magenta';

window.addEventListener('DOMContentLoaded', (event) => {
  const svg = document.querySelector('.inner svg')
  svg.parentNode.replaceChild(paper.project.exportSVG(), svg);
});

// existing code...

export default Pattern
```

This code defines two paths, and executes a boolean operation on the `bodiceFrontSource` (which is just a triangle here) -- subtracting the arm hole shape (a simple circle) to end up with the `bodiceFront` path.

This means that things like armhole shapes can be calculated as full shapes, and can then be subtracted from the pattern part in question. I envision creating a couple base modules, like `armHole`, `neckHole`, `dart` and potentially an `advancedDart` (with more advanced curves that could be used on seams like the shoulder seam and side seam -- which generally function the same way darts do).

Screenshot:

<img width="554" alt="Screen Shot 2021-06-12 at 1 36 48 PM" src="https://user-images.githubusercontent.com/13765/121788291-5c093f00-cb89-11eb-8c74-a1a6b71082d0.png">

I don't know if this gets us to a completely modular concept, but it might help?

**Comment by joostdecock** [comment]
> would you be willing to put together a quick summary of what Part expects of the instances inside the Part.paths object?

This image [from our docs](https://freesewing.dev/guides/patterns/) is a good starting point:

![Screenshot 2021-06-13 at 09-39-07 Pattern guide](https://user-images.githubusercontent.com/1708494/121799236-42b8cf00-cc2b-11eb-83fd-612f3beb2671.png)

A `part` is a container for a bunch of data types that are used to draft a pattern, specifically points, paths, and snippets.

 - A point is straight forward, it has a `x` and `y` property, storing its coordinates in a 2D plane.
 - A snippet is used for the SVG `use` tag, it has a `def` property holding the reference to the snippet's definition in the SVG `defs` section, and an `anchor` property that holds a point
 - A path is how we store data for lines/curves. Its `ops` property holds an array of drawing operations. Each is an object as such:
 ```
 { 
   type: [move|line|curve],
   to: point,
   cp1: point (only used by curves)
   cp2: point (only used by curves)
}
```
When it's time to render, we construct the SVG pathstring by glueing these different ops together.

So as a way to store data, these are all rather straight-forward. It's the methods that come with these objects that provide the added value.

**Reply by jejacks0n** [reply]
🙏 thank you for this write up @joostdecock!

**Comment by BenJamesBen** [comment]
I want a Part to tell me what it is and what it provides me. 
- Hello, I am a Brian bodice front part. 
- I am symmetrical, to be cut on a fold.
- I provide you with 4 places to attach additional parts: at the neckline, armsceye (2 of them), side (2 of them), and waist. 
- On the sides, you can attach a bodice back part. At the armsceyes, you can attach a sleeve part. At the neckline, you can attach a collar.
- I provide the following official Points: (list of all point names and what they represent).
- I provide the following official Paths: (list of all paths and what they represent).
- I provide the following Notches: (list of all notches and what they represent).
- I need the following mandatory and optional measurements: (list of measurements).
- I make the following assumptions about measurements: (list of assumptions about measurement ranges and relationships.)
- I provide A, B, and C amounts of ease in sections X, Y, and Z. 
- I am compatible with the following Designs: (list of Designs).
- I am compatible with the following Parts: (list of Parts).
- I provide you with the amount of material needed to cut me.

By "tell me" I mean the Part has a way to communicate these things to me-- the end customer. If the same method can be used to communicate these things with other Parts and Designs (so you could automagically combine different parts together to make new designs), that would be a bonus. However, the Part first needs to communicate these things to the customer. That is the most important thing.

Most of these are things that our Parts and Designs currently do not tell us. (We can find out names of individual elements by digging into the code and using debug/xray mode, but we do not know what these elements are intended for and which ones should be used by other parts/designs. Sometimes there are notches, and it is unclear what they are intended for.)

**Comment by joostdecock** [comment]
This is stale and also never really grew legs. Closing.
